### PR TITLE
New sorting order dialog

### PR DESCRIPTION
**Pull request by AndyScherzinger**
PR for #340

device-2017-01-26-125301

### CODE REVIEW

**Comment by mario** [comment]
Do we even need a cancel? If it's a dialog, any click outside the dialog should cancel it.

**Comment by AndyScherzinger** [comment]
@mario that is correct (and would also work), so we could remove it. 

What do you think @eppfel ?

**Comment by mario** [comment]
@AndyScherzinger how do you currently show the selected/default choice? That's not visible in the screenshot you've shown us.

**Comment by eppfel** [comment]
How could we make modification time and size more understandable? The arrow in the icon not self-explanatory IMO. 🤔 

> Do we even need a cancel? If it's a dialog, any click outside the dialog should cancel it.

If it does not make the pop-up overflow, I would leave it. Gives the insecure user more control. Worst case: some user chooses sorting out of accident and does not know, he can tap on the background to cancel.

> @AndyScherzinger how do you currently show the selected/default choice? That's not visible in the screenshot you've shown us.

I think he wanted to highlight the icon with the nextcloud/theme color. 👍

**Comment by AndyScherzinger** [comment]
> How could we make modification time and size more understandable? The arrow in the icon is not self-explanatory IMO.

I just created the icons to have something, happy for any contributions for iconswork :)



> @AndyScherzinger how do you currently show the selected/default choice? That's not visible in the screenshot you've shown us.

> I think he wanted to highlight the icon with the nextcloud/theme color. 

exactly, but that part isn't finished yet, the active one will be tinted with the theme color 😃

**Comment by eppfel** [comment]
> I just created the icons to have something, happy for any contributions for iconswork :)

 I thought there are defautls in the material icons set. 😁

**Comment by AndyScherzinger** [comment]
> I thought there are defautls in the material icons set.

There are non for this, so I created them based on existing material design icons (basically combining the "clock" and "space" with the arrow...)

**Comment by eppfel** [comment]
For size we could use this: [sortic_sort_black_24dp_1x](https://material.io/icons/#ic_sort)  and upside down, instead of the arrow.

Does not make too much sense for time, though...

**Comment by eppfel** [comment]
And "Z over A" would be awesome for the second icon.

**Comment by AndyScherzinger** [comment]
> And "Z over A" would be awesome for the second icon.

And the arrow? would that one than have to point down for both?

> For size we could use this: [sortic_sort_black_24dp_1x](https://material.io/icons/#ic_sort)   and upside down, instead of the arrow.

while keeping the "cake" part?

**Comment by AndyScherzinger** [comment]
Screenshot with the tinting active:
device-2017-01-23-125809

**Comment by eppfel** [comment]
>> And "Z over A" would be awesome for the second icon.
>
>And the arrow? would that one than have to point down for both?

mh, logically, yes. I guess I have to see to see it. 

>> For size we could use this: [sortic_sort_black_24dp_1x](https://material.io/icons/#ic_sort)   and upside down, instead of the arrow.
>
>while keeping the "cake" part?

Yes

Hope you're having fun with the icons.. 😆
Font awesome has some different sorting icons: http://fontawesome.io/icons/ -> search "sort"

**Comment by AndyScherzinger** [comment]
> mh, logically, yes. I guess I have to see to see it.

see screenshot above and below the latest one with the changed size icon:
device-2017-01-23-145201

**Comment by eppfel** [comment]
* arrow on A->Z has to be upside down
* I would put bigger -> smaller left and smaller -> bigger 
* make the sort icon bigger in relevance to the cake

**Comment by AndyScherzinger** [comment]
> I would put bigger -> smaller left and smaller -> bigger

What would be "bigger" for all three types? (size is the one I understand)

**Comment by tobiasKaminsky** [comment]
maybe it is just my lazy brain...but modification time not clear to me...left is "newest -> oldest"?
maybe something like a triangle is better for this (instead of an arrow)?

**Comment by AndyScherzinger** [comment]
Anybody is welcome to do the icons ;) cc: @jancborchardt 

I just put some effort into it to get things started, but I am not *-designer, just want to get this thing done! 🚀

**Comment by eppfel** [comment]
bigger menat only file-size

**Comment by AndyScherzinger** [comment]
@eppfel you mean like this?

|Name                        |A->Z                 |Z->A                  |
|Modification date     |younger->older|older->younger|
|Size                           |bigger->smaller|smaller->bigger|

And if so, how should the modification date icons be changed? With the triangles like @tobiasKaminsky suggested? And if so is triangle "up" then younger->older or the other way round?

**Comment by eppfel** [comment]
Layout 👍 

> how should the modification date icons be changed? With the triangles like @tobiasKaminsky suggested? And if so is triangle "up" then younger->older or the other way round?

I would use the sort icon, too: shorter bars = shorter time?

**Comment by AndyScherzinger** [comment]
So here is the latest version @eppfel:
device-2017-01-24-224718

**Comment by jancborchardt** [comment]
Sorry to interject. But except the A-Z and Z-A, these icons are not understandable.

The table in @AndyScherzinger's comment is the best UI actually: https://github.com/nextcloud/android/pull/575#issuecomment-274739957

We have 3*2 options, with the left one always being the default:
- A -> Z / Z -> A
- Newest first / Oldest first
- Biggest first / Smallest first

And we should use these words (not ascending/descending cause they are not understandable).

Even if it's just an entry saying "Sort by newest first" and tap on it opens a radio list with 6 entries. As words, not icons.

**Comment by AndyScherzinger** [comment]
Okay, not pushing the latest commit then.

So let me be frustrated here... This just dumps hours of work obsolete simply to end up with a solution that would it have been suggested in the first place would have taken 5 (!) minutes to implement! 😠

Other than that the proposal is confusing me. I would click "sort" in the menu and then the dialog comes up and then I choose (atm 3 opens), then 6 (like mentioned by Jan).
In doing so (not adding a cancel button), one would simply revert the complete PR, extend the corresponding string array and update the switch/case logic in File Listing and Upload Listing activities. Done!

So feel free to take over!

**Comment by eppfel** [comment]
@jancborchardt Yes, the icons are hard to read and your solution solves this easily.

@AndyScherzinger Sry, that I did think of it before. But please don't throw it away completely. I think , if we will add more options the list will get too long. And we could introduce the icons with @jancborchardt's solution as well, because after some time users are able to learn them.

**Comment by AndyScherzinger** [comment]
@eppfel I didn't, the branch is still there, I just closed the actual PR.

So the question would now be how it should be and then the implementation can be done. As in should it then simply be the radio button list with the six options?

**Comment by jancborchardt** [comment]
Sorry, this is entirely my fault for not seeing this earlier and commenting earlier.

Here is a quick mockup with an improvement of the current dialog which includes the new sort options and the nice icons (which do work _with_ text):
sorting
Additional considerations:
- Put a small bit of whitespace between the 3 sets of two related options to group them, if that’s possible
- I’m not particular about the cancel button. But I would say we should keep it because it might not be entirely obvious that clicking outside (or clicking the selected sort option again) will close the popover. We can also remove it though if you think that’s fine.

The important thing is that this dialog will be used very rarely, so it has to be understandable with 0 thinking. Relying on a matrix of icons doesn’t work for that.

What do you think @eppfel @AndyScherzinger @mario?

**Comment by mario** [comment]
Looks nice.

**Comment by AndyScherzinger** [comment]
Looks awesome! :) (even though the icons aren't great but work...I am no icon designer, just did my best...).
I'll implement the change, probably later today :)  🚀 

> Put a small bit of whitespace between the 3 sets of two related options to group them, if that’s possible with Android. If not, nevermind.

consider it done! 👍 

> I’m not particular about the cancel button. But I would say we should keep it because it might not be entirely obvious that clicking outside (or clicking the selected sort option again) will close the popover. We can also remove it though if you think that’s fine.

It is not part of the actual/1.4.0 UI (3 radio buttons) but we discussed if we should introduce it. I just checked GMail and they do show a cancel button for their chooser/selection dialogs, so I#d say we are in line (and yes back-button, clicking next to the pop-up, click on the already active item also closes the screen):
device-2017-01-25-132923

Thanks for the nicer solution to the UI/UX problem @jancborchardt And also on the plus side this way nearly the complete implementation I've done can be used, I just need to adapt the slightly changes layout 🎉

**Comment by jancborchardt** [comment]
I have nothing more to add than :heart: :)

**Comment by AndyScherzinger** [comment]
Then here is the commit https://github.com/nextcloud/android/pull/575/commits/d9ee50764edf9d96691e7687da602d6fc468c446 and the screenshot:
device-2017-01-25-154816

**Comment by tobiasKaminsky** [comment]
As it is ready for review can you remove the WIP?

**Comment by AndyScherzinger** [comment]
@mario can you also review? final screenshot is https://github.com/nextcloud/android/pull/575#issuecomment-275128290

**Comment by ldmpub** [comment]
Hi All,

I'm glad to see the final choice https://github.com/nextcloud/android/pull/575#issuecomment-275128290

Thank's to all and special thanks to @AndyScherzinger, I'm eager to have it in hand with next release!

**Comment by AndyScherzinger** [comment]
@ldmpub thanks happy you like it. We are now targeting 1.4.2 or 1.5.0 since the next release 1.4.1 is already done and the stable will be released probably by the end of this week :)

**Comment by AndyScherzinger** [comment]
@eppfel @jancborchardt what do you think of the latest screenshot?  I am not sure if if should put more whitespace between icon and text...

**Comment by mario** [comment]
@AndyScherzinger 

a) Please don't add any more spacing
b) I'd target this for 1.5.0 as it's a UX change, and not a bugfix

**Comment by mario** [comment]
Gonna review it now.

**Comment by AndyScherzinger** [comment]
Thanks @mario - 1.5.0 sounds good to me 👍

**Comment by jancborchardt** [comment]
Great! Also on the spacing and everything! Just one thing: To accomodate people with difficulties seeing color, can we also bold the typeface of the currently selected elemwnt?

**Comment by AndyScherzinger** [comment]
@jancborchardt Done: https://github.com/nextcloud/android/pull/575/commits/1260b12f80fae14becf662c8bec99fbcb337bbda
device-2017-01-26-125301

**Comment by AndyScherzinger** [comment]
This might solve https://github.com/nextcloud/android/issues/572 - So I'd say we also release it with 1.4.2

What do you think @mario @tobiasKaminsky

**Comment by mario** [comment]
@AndyScherzinger all good but since you can't reproduce ... 1.5.0 :)

**Comment by AndyScherzinger** [comment]
@mario 😃 👍

**Comment by AndyScherzinger** [comment]
@mario @tobiasKaminsky can you do the code review? :)

**Comment by mario** [comment]
@AndyScherzinger certainly, but this is 1.5.0 - can you please review the filtering improvements which is 1.4.2 and check that Auto upload PR which is also 1.4.2? :)

File: res/layout/sorting_order_fragment.xml [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="0dp"
+        android:layout_weight="1">
+
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="vertical">
+
+            <LinearLayout
```
Comment (mario) [inline res/layout/sorting_order_fragment.xml]: Can you please clarify why we have so many linear layouts inside linear layouts? That takes a while to render the way it's set in this layout file.

Comment (AndyScherzinger) [inline res/layout/sorting_order_fragment.xml]: No real reason, just forget to do the housekeeping, replace it with a table layout via https://github.com/nextcloud/android/pull/575/commits/f39c2174470762bad146fb7dbb446ee8894b0891

**Comment by AndyScherzinger** [comment]
rebased and can now also be merged to master since 1.4.2 has been released. @mario @tobiasKaminsky please approve/review

**Comment by AndyScherzinger** [comment]
@jamiehutber just wanted to let you know that we added advanced sorting options with this PR that will be part of our next feature release.

**Comment by jamiehutber** [comment]
Nice! Cheers guys

Thanks
Jamie

**Comment by AndyScherzinger** [comment]
@jamiehutber just to make sure you caught this in the notification mail this is the Nextcloud Android client (compatible with ownCloud) ;)

**Comment by jamiehutber** [comment]
Ah, I did not know that. I will be sure to check it out in that case 👍

**Comment by AndyScherzinger** [comment]
@mario @tobiasKaminsky your turn :) Can imho also be merged since 1.4.2 has been released :)

File: src/com/owncloud/android/ui/activity/UploadFilesActivity.java [hunk 2]
```diff
@@ -246,6 +241,33 @@ public void onClick(DialogInterface dialog, int which) {
...
+                break;
+            case SortingOrderDialogFragment.BY_SIZE_DESC:
+                mFileListFragment.sortBySize(false);
+                break;
+            default: // defaulting to alphabetical-ascending
+                Log_OC.w(TAG, "Unknown sort order, defaulting to alphabetical-ascending!");
+                mFileListFragment.sortByName(true);
+                break;
+        }
+    }
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/ui/activity/UploadFilesActivity.java]: Too much headache, but can you extract this part to an external function and re-use it? Cause now it is two times exact the same code @AndyScherzinger

**Comment by AndyScherzinger** [comment]
@mario I also rebased it to latest master, so can we also try to get this back to master, might do a final test today and would then be happy if it can be merged and you could take a final look at the PR.

**Comment by AndyScherzinger** [comment]
@mario rebased to latest master and successfully tested it, so I leave the merge to you if you are fine with the change (I know you already approved the PR, but I rebased).

**Comment by mario** [comment]
Merged, thank you for your work.

**Pull request merged**